pkg/apis/configuration/v1alpha1/kong_vault_types.go

```diff
@@ -0,0 +1,85 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	KongValtKind = "KongVault"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:object:root=true
+// +kubebuilder:storageversion
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:scope=Cluster,shortName=kv,categories=kong-ingress-controller
+// +kubebuilder:resource:path=kongvaults
+// +kubebuilder:printcolumn:name="Backend-Type",type=string,JSONPath=`.spec.backend`,description="Name of the backend of the vault"
+// +kubebuilder:printcolumn:name="Prefix",type=string,JSONPath=`.spec.prefix`,description="Prefix of vault URI to reference the values in the vault"
+// +kubebuilder:printcolumn:name="Age",type=date,JSONPath=`.metadata.creationTimestamp`,description="Age"
+// +kubebuilder:printcolumn:name="Description",type=string,JSONPath=`.spec.description`,description="Description",priority=1
+// +kubebuilder:printcolumn:name="Programmed",type=string,JSONPath=`.status.conditions[?(@.type=="Programmed")].status`
+
+type KongVault struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec              KongVaultSpec   `json:"spec"`
+	Status            KongVaultStatus `json:"status,omitempty"`
+}
+
+type KongVaultSpec struct {
+	// Backend is the type of the backend storing the values in the vault.
+	// +kubebuilder:validation:MinLength=1
+	Backend string `json:"backend"`
+	// Prefix is the prefix of vault URI for referencing values in the vault.
+	// +kubebuilder:validation:MinLength=1
+	Prefix string `json:"prefix"`
+	// Description is the description to tell the information about the vault.
+	Description string `json:"description,omitempty"`
+	// Config is the configuration of the vault. Varies for different backends.
+	Config apiextensionsv1.JSON `json:"config,omitempty"`
+}
+
+type KongVaultStatus struct {
+	// Conditions describe the current conditions of the KongVaultStatus.
+	//
+	// Known condition types are:
+	//
+	// * "Programmed"
+	//
+	// +listType=map
+	// +listMapKey=type
+	// +kubebuilder:validation:MaxItems=8
+	// +kubebuilder:default={{type: "Programmed", status: "Unknown", reason:"Pending", message:"Waiting for controller", lastTransitionTime: "1970-01-01T00:00:00Z"}}
+	Conditions []metav1.Condition `json:"conditions"`
+}
+
+// +kubebuilder:object:root=true
+
```

Comment:
```suggestion
// KongVaultList contains a list of KongVault.
// +kubebuilder:object:root=true
```
We have it like that in other CRDs